spouse/spouse_demo.py

```diff
@@ -0,0 +1,552 @@
+#
+# <img align="left" src="imgs/marginals-common.jpg" width="320px" style="margin-left:0px">
+#
+# <img align="left" src="imgs/marginals-real.jpg" width="320px" style="margin-right:0px">
+#
+# <img align="right" src="imgs/marginals-ideal.jpg" width="320px" style="margin-right:0px">
+
+# %%
+import matplotlib.pyplot as plt
+
+train_proba = label_model.predict_proba(train_L)
+plt.hist(train_proba[:, 1], bins=20, range=(0.0, 1.0))
+plt.show()
+
+# %% [markdown]
+# ## Part 4: Training our End Extraction Model
+#
+# In this final section of the tutorial, we'll use our noisy training labels alongside the development set labels to train our end machine learning model. We start by filtering out training examples which did not recieve a label from any LF, as these examples contain no signal. Then we concatenate them with dev set examples.
+#
+# %%
+import pandas as pd
+from snorkel.analysis.utils import preds_to_probs
+
+dev_label_probs = preds_to_probs(dev_labels, 2)
+mask = train_L.max(1) != -1
+combined_df = pd.concat([dev_df, train_df.iloc[mask]])
+combined_proba = np.concatenate([dev_label_probs, train_proba[mask, :]], 0)
+
+# %% [markdown]
+# ## II. Training a _Long Short-term Memory_ (LSTM) Neural Network
+#
+# [LSTMs](https://en.wikipedia.org/wiki/Long_short-term_memory) can acheive state-of-the-art performance on many text classification tasks. We'll train a simple LSTM model below. tf_model contains functions for processing features and building the tensorflow graphs for training and evaluation.
+#
+# For this tutorial, we will be training a fairly effective deep learning model. More generally, however, Snorkel plugs in with many ML libraries, making it easy to use almost any state-of-the-art model as the end model!
+
+# %%
+from tf_model import get_features_and_labels, get_model
+
+model = get_model()
+tokens, idx1, idx2, label_probs = get_features_and_labels(
+    combined_df, combined_proba, tf.float32
+)
+batch_size = 64
+num_epochs = 2  # TODO: Change this to ~10. Warning: Training takes tens of minutes!
+model.fit(
+    (tokens, idx1, idx2),
+    label_probs,
+    batch_size=batch_size,
+    steps_per_epoch=(len(train_df) // batch_size),
+    epochs=num_epochs,
+)
+
+
+# %% [markdown]
+# Measure the trained model's prediction accuracy.
+
+# %%
+# Truncate sentences to limit memory usage when padding.
+def get_feature_arrays(df):
+    def pad_or_truncate(l, max_length=60):
+        return l[:max_length] + [""] * (max_length - len(l))
+
+    tokens = np.array(list(map(pad_or_truncate, df.tokens)))
+    idx1 = np.array(list(map(list, df.person1_word_idx)))
+    idx2 = np.array(list(map(list, df.person2_word_idx)))
+    return tokens, idx1, idx2
+
+
+test_tokens, test_idx1, test_idx2 = get_feature_arrays(test_df)
+probs = model.predict((test_tokens, test_idx1, test_idx2))
+preds = probs_to_preds(probs)
+print(
+    f"Test Accuracy when trained with soft labels: {metric_score(test_labels, preds=preds, metric='accuracy')}"
```

Comment:
f1 instead of accuracy

Reply:
F1 score is always zero, because the model always assigns more than 0.5 probability to negative (due to the dataset imbalance). So we use ROC-AUC to measure model quality (the baseline that always outputs negative gets ROC-AUC of 0.5).

Reply:
TODO @paroma to tune LabelModel to fix this